File: Stack/leetcode_32.py
# Given a string containing just the characters '(' and ')', return the length of the longest valid(well - formed) parentheses substring.

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]  # Initialize the stack with -1 to handle edge cases
        max_len = 0

        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)  # Push the index of '(' onto the stack
            else:
                stack.pop()  # Pop the '(' index from the stack

                if not stack:
                    # If the stack is empty, push the current index
                    stack.append(i)
                else:
                    # Calculate the length of valid parentheses
                    max_len = max(max_len, i - stack[-1])

        return max_len


# Test the Solution class
sol = Solution()
input_str = "(()))())("  # Example input
result = sol.longestValidParentheses(input_str)
print("Length of the longest valid parentheses:", result)

File: Stack & Queue/leetcode_32.cpp

// Problem Link ====>>https://leetcode.com/problems/longest-valid-parentheses/description/

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int longestValidParentheses(string s)
    {
        stack<int> st;
        st.push(-1); // Initialize the stack with -1 to handle edge cases

        int maxLen = 0;
        for (int i = 0; i < s.length(); ++i)
        {
            if (s[i] == '(')
            {
                st.push(i); // Push the index of '(' onto the stack
            }
            else
            {
                st.pop(); // Pop the '(' index from the stack

                if (st.empty())
                {
                    st.push(i); // If the stack is empty, push the current index
                }
                else
                {
                    maxLen = max(maxLen, i - st.top()); // Calculate the length of valid parentheses
                }
            }
        }

        return maxLen;
    }
};

int main()
{
    Solution sol;               // Create an instance of the Solution class
    string input = "(()))())("; // Example input
    int result = sol.longestValidParentheses(input);
    cout << "Length of the longest valid parentheses: " << result << endl;

    return 0;
}